ORC-611: Min/max column statistics precision fix

pgaref opened the pull request:
Column statistics have ms precision, while timestamps come with ns precision,
We should round-down min and round-up max column statistics to make sure PPD with ns precision works

Change-Id: I22e89b102ffc3aa9750a30bbfce5c847993bf927

pgaref commented:
> Thanks for the quick fix!
> I didn't read through the tests carefully yet, I will do that later.

Hey @csringhofer -- thanks a lot for the comments!
I put some more thought into the patch and it seems that we can sort out the issue on the Reader side avoiding changing the Column stats that are written and having to support another writer version.

In the latest patch, whenever we are evaluating a predicate with Timestamps, I just widen the min/max range by 1 ms to make sure the underlying ns range is included -- this should solve the issue both on Java and C++ consumers using our Readers  (until we feel comfortable changing the column statistics on the writer side and remove this change on the reader).

Thoughts?

omalley commented on java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:
```diff
@@ -405,7 +405,12 @@ static ValueRange getValueRange(ColumnStatistics index,
       Timestamp min = useUTCTimestamp ? stats.getMinimumUTC() : stats.getMinimum();
       Timestamp max = useUTCTimestamp ? stats.getMaximumUTC() : stats.getMaximum();
-      return new ValueRange<>(predicate, min, max, stats.hasNull());
+      // To support predicates with ns precision we widen min and max ms range by 1 to ensure ns values are in range.
```
This is the wrong pattern, unfortunately. You'll need to follow the example of what was done for the string shortening. (We just store the first 1024 characters for min/max on strings.) The problem comes that if someone does a test that looks for equality, this change will return wrong results.

pgaref replied:
Thanks for the comments @omalley ! Latest patch updates TimestampStatistics to store the lower 6 digits (that were initially ignored) in order to achieve nanosecond precision.

Also liked the idea proposed by @csringhofer and added 0 and 999_999 as DEFAULT_NANOS MIN and MAX respectively when these values are missing.

Please let me know what you think!

omalley commented on java/core/src/test/org/apache/orc/TestOrcTimestampPPD.java:
```diff
@@ -0,0 +1,173 @@
+        SearchArgumentFactory.newBuilder()
+            .equals("c", PredicateLeaf.Type.TIMESTAMP, Timestamp.valueOf("2037-01-01 00:00:00.001009")).build()
+            .getLeaves().get(0), null));
+  }
```
Please add a test with a timestamp equality and both approximate and precise min/max.

csringhofer commented on c++/src/sargs/PredicateLeaf.cc:
```diff
@@ -536,11 +536,17 @@ namespace orc {
             literal2Timestamp(mLiterals),
-            stats.minimumutc(),
-            stats.maximumutc(),
+            minimum-1,
```
Decreasing the minimum is probably needed, as the writers round towards negative infinity.

csringhofer commented on java/core/src/test/org/apache/orc/TestOrcTimezonePPD.java:
```diff
@@ -294,7 +294,7 @@ public void testTimestampMinMaxAndBloomFilter() throws Exception {
     assertEquals(writerTimeZone, TimeZone.getDefault().getID());
     List<String> ts = Lists.newArrayList();
-    ts.add("2007-08-01 00:00:00.0");
+    ts.add("2007-08-01 01:00:00.0");
```
What is the reason behind this 1 hour change? It doesn't seem relevant to sub-millisec issues.

csringhofer commented on java/core/src/test/org/apache/orc/TestOrcTimestampPPD.java:
```diff
@@ -0,0 +1,173 @@
+    Timestamp gotMin = ((TimestampColumnStatistics) colStats[0]).getMinimum();
+    assertEquals("1970-01-01 00:00:00.0", gotMin.toString());
+
+    Timestamp gotMax = ((TimestampColumnStatistics) colStats[0]).getMaximum();
```
A comment could explain that this is a bit buggy in the writer, but the reader can handle it.

pgaref replied:
A proper solution, in general, would include lower and upper bounds similar to what we do for string shortening (https://issues.apache.org/jira/browse/ORC-203) -- see Owens reply above.

To achieve that we cannot really avoid bumping the writer version so I will update the patch accordingly

pgaref replied:
@csringhofer please let me know what you think about the latest changes!

csringhofer commented on java/core/src/test/org/apache/orc/TestOrcTimestampPPD.java:
```diff
@@ -0,0 +1,173 @@
+    while (rows.nextBatch(batch)) {
+      for (int r = 0; r < batch.size; ++r) {
+        Assert.assertEquals(tslist.get(0), times.asScratchTimestamp(r));
+        Assert.assertEquals(tslist.get(0).getNanos(), times.asScratchTimestamp(r).getNanos());
```
I don't get this - why are the nanos always equal with the first element in 'tslist'? Is it always 0?

pgaref replied:
The reason is all **tslist** Timestamps have the same value up to the supported MS precision -- just made this assumption for the test

pgaref commented on java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java:
```diff
@@ -1815,7 +1893,7 @@ public int hashCode() {
     @Override
     public void updateTimestamp(Timestamp value) {
-      updateTimestamp(value.getTime());
+      updateTimestamp(value.getTime(), value.getNanos());
```
correct, this is now fixed

pgaref commented on java/core/src/java/org/apache/orc/impl/writer/TimestampTreeWriter.java:
```diff
@@ -111,7 +111,7 @@ public void writeBatch(ColumnVector vector, int offset,
         }
```
Makes sense to me, thats why in getMinimum and getMaximum methods in TimestampStatistics I set the nanos after the utc conversion

pgaref commented on java/core/src/test/org/apache/orc/TestOrcTimestampPPD.java:
```diff
@@ -0,0 +1,208 @@
+
+    // PPD Equals with ms precision NOT passing
+    pred = createPredicateLeaf(PredicateLeaf.Operator.EQUALS, PredicateLeaf.Type.TIMESTAMP, "c",
+        Timestamp.valueOf("2037-01-01 00:00:00"), null);
```
Thanks, fixed

pgaref commented on java/core/src/test/org/apache/orc/TestOrcTimestampPPD.java:
```diff
@@ -0,0 +1,208 @@
+        Timestamp.valueOf("2037-01-01 00:00:00"), null);
+    Assert.assertEquals(SearchArgument.TruthValue.NO, RecordReaderImpl.evaluatePredicate(colStats[0], pred, null));
+
+    // PPD Equals with ns precision passing
```
fixed

pgaref commented on java/core/src/test/org/apache/orc/TestOrcTimestampPPD.java:
```diff
@@ -0,0 +1,208 @@
+        Timestamp.valueOf("2037-01-01 00:00:00.006789"), null);
+    Assert.assertEquals(SearchArgument.TruthValue.YES_NO, RecordReaderImpl.evaluatePredicate(colStats[0], pred, null));
+
+    // PPD Equals with ms precision passing
```
fixed

wgtmac commented on java/core/src/java/org/apache/orc/TimestampColumnStatistics.java:
```diff
@@ -47,4 +47,19 @@
+   * The minimum value is truncated at ms scale.
+   * @return the 6 lowest digits of minimum value
+   */
+  int getMinimumNanos();
```
java.sql.Timestamp does support nanosecond. Is it better to use existing APIs and handle nanosecond transparently?

pgaref replied:
We could do that, however I believe its more clear to handle the nanoseconds as separate methods to show how ns precision is handled in old Proto (default values) and new Proto with existing values.

wgtmac commented on java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java:
```diff
@@ -1670,46 +1675,88 @@ public int hashCode() {
 
     @Override
-    public void updateTimestamp(long value) {
+    public void updateTimestamp(long value, int nanos) {
```
Just curious, can we directly use java.sql.Timestamp to store minimum & maximum in UTC? If that is true, we can deprecate updateTimestamp and make the comparison and merge process easier.

pgaref replied:
That would be useful, havent checked but I believe this could be part of a followup ticket.

wgtmac commented on java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java:
```diff
@@ -1670,46 +1675,88 @@ public int hashCode() {
+            maxNanos = timestampStats.maxNanos;
+          } else {
+            /* both are empty */
+            maximum = minimum = null;
```
the else branch can be deleted

pgaref replied:
Makes sense, thanks

omalley commented:
Sorry for the long delay on this one. I made some changes that should be looked over. I pushed my changes to my [orc-611 branch](https://github.com/omalley/orc/tree/orc-611). The main one was changing the protobuf to used unsigned ints for min & max nanos. To ensure that we don't lose the ability to detect the old files (when we move to protobuf 3), I added one to the values so that the range of both fields is 1 to 1_000_000. (Protobuf 3 doesn't serialize the default value for fields, so a 0 would not be serialized.)

omalley commented:
@pgaref Can you take a look at my change to see if it looks reasonable?

pgaref commented:
> @pgaref Can you take a look at my change to see if it looks reasonable?

Hey @omalley thanks for the comment -- happy to take a look just give me a couple of days as I am currently OOO

omalley commented:
@pgaref when you get a chance, it would be great.

pgaref commented:
Hey @omalley just did a bit of polishing and squashed everything.
Thanks for the protobuf suggestion!  It does make sense to me and it also simplified the min/max calculation as a side effect! Latest patch LGTM
